### ISSUE TITLE
502 Bad Gateway on work-pools page behind a nginx reverse proxy

### ISSUE THREAD

**Issue by andreadistefano** [PrefectHQ/prefect#8918]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

Hi, I just started experimenting with prefect and everything looks really cool so far, so first of all thank you for all the work you're doing! Also, English is not my first language, but I hope my message is clear enough.

I'm trying to run the prefect server on a remote Linux machine (not using Docker, just `pip install` inside a python 3.11 virtualenv) and run the flows on that machine.

The remote prefect server is behind a reverse proxy at the address https://prefect.mycompany.tld, so I configured the `PREFECT_API_URL` and `PREFECT_UI_API_URL` environment variables accordingly.

I successfully ran some flows on my Windows machine and they showed up in the web UI, so everything was going great.
Then, I set up an agent on the same Linux machine with the command `prefect agent start -p "default-agent-pool"`, using the default pool I saw on the web UI.

Then, I ran a simple deployment from my machine, which ran successfully and showed up in the UI, but now the prefect server is acting up:

- randomly getting a 502 Bad Gateway error from nginx on https://prefect.mycompany.tld/work-pools (sometimes it loads fine)
- when the work-pools page loads, the queues are reported as Unhealty
- `prefect work-pool ls` gives the error in the section below
- `prefect version` reports `<client error>` in `Server type` field

I really don't know what I've done wrong, but I'll be happy to give more feedback/logs/anything if someone directs me in how to get them.
EDIT: I can also "reset" the server if needed, as there is literally nothing on it at the moment.

### Reproduction

```python3
# Sorry guys, I don't know what to say. Here is the deployment I ran but I don't really think the problem is here

from main import import_excel_flow
from prefect.deployments import Deployment

deployment = Deployment.build_from_flow(
    flow=import_excel_flow,
    name="test-deployment",
    version=1,
    work_queue_name="demo",
)
deployment.apply()
```

### Error

```python3
Traceback (most recent call last):
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/api/server.py", line 453, in lifespan
    await run_migrations()
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/api/server.py", line 376, in run_migrations
    await db.create_db()
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/database/interface.py", line 55, in create_db
    await self.run_migrations_upgrade()
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/database/interface.py", line 63, in run_migrations_upgrade
    await run_sync_in_worker_thread(alembic_upgrade)
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/utilities/asyncutils.py", line 91, in run_sync_in_worker_thread
    return await anyio.to_thread.run_sync(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
           ^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/database/alembic_commands.py", line 24, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/database/alembic_commands.py", line 53, in alembic_upgrade
    alembic.command.upgrade(alembic_config(), revision, sql=dry_run)
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/alembic/command.py", line 378, in upgrade
    script.run_env()
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/alembic/script/base.py", line 576, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/alembic/util/pyfiles.py", line 94, in load_python_file
    module = load_module_py(module_id, path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/alembic/util/pyfiles.py", line 110, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/database/migrations/env.py", line 147, in <module>
    apply_migrations()
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/utilities/asyncutils.py", line 248, in coroutine_wrapper
    return run_async_from_worker_thread(async_fn, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/utilities/asyncutils.py", line 177, in run_async_from_worker_thread
    return anyio.from_thread.run(call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/anyio/from_thread.py", line 49, in run
    return asynclib.run_async_from_thread(func, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 970, in run_async_from_thread
    return f.result()
           ^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/database/migrations/env.py", line 137, in apply_migrations
    engine = await db_interface.engine()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/database/interface.py", line 73, in engine
    engine = await self.database_config.engine()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/database/configurations.py", line 187, in engine
    raise RuntimeError(
RuntimeError: Prefect requires sqlite >= 3.24.0 but we found version 3.7.17

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/starlette/datastructures.py", line 702, in __getattr__
    return self._state[key]
           ~~~~~~~~~~~^^^^^
KeyError: 'services'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/cli/_utilities.py", line 41, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/utilities/asyncutils.py", line 260, in coroutine_wrapper
    return call()
           ^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/_internal/concurrency/calls.py", line 245, in __call__
    return self.result()
           ^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/_internal/concurrency/calls.py", line 173, in result
    return self.future.result(timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/_internal/concurrency/calls.py", line 218, in _run_async
    result = await coro
             ^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/cli/work_pool.py", line 93, in ls
    async with get_client() as client:
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/client/orchestration.py", line 2301, in __aenter__
    self._ephemeral_lifespan = await self._exit_stack.enter_async_context(
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/contextlib.py", line 638, in enter_async_context
    result = await _enter(cm)
             ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/contextlib.py", line 204, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/client/base.py", line 92, in app_lifespan_context
    await context.__aenter__()
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/asgi_lifespan/_manager.py", line 92, in __aenter__
    await self._exit_stack.aclose()
  File "/usr/local/lib/python3.11/contextlib.py", line 675, in aclose
    await self.__aexit__(None, None, None)
  File "/usr/local/lib/python3.11/contextlib.py", line 733, in __aexit__
    raise exc_details[1]
  File "/usr/local/lib/python3.11/contextlib.py", line 716, in __aexit__
    cb_suppress = await cb(*exc_details)
                  ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/asgi_lifespan/_concurrency/asyncio.py", line 80, in __aexit__
    await self.task
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/asgi_lifespan/_concurrency/asyncio.py", line 63, in run_and_silence_cancelled
    await self.coroutine()
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/asgi_lifespan/_manager.py", line 64, in run_app
    await self.app(scope, self.receive, self.send)
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/fastapi/applications.py", line 276, in __call__
    await super().__call__(scope, receive, send)
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/starlette/applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/starlette/middleware/errors.py", line 149, in __call__
    await self.app(scope, receive, send)
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/api/server.py", line 114, in __call__
    await self.app(scope, receive, send)
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/starlette/middleware/cors.py", line 76, in __call__
    await self.app(scope, receive, send)
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 55, in __call__
    await self.app(scope, receive, send)
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/starlette/routing.py", line 707, in __call__
    await self.lifespan(scope, receive, send)
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/starlette/routing.py", line 677, in lifespan
    async with self.lifespan_context(app) as maybe_state:
  File "/usr/local/lib/python3.11/contextlib.py", line 204, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/api/server.py", line 458, in lifespan
    await stop_services()
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/prefect/server/api/server.py", line 440, in stop_services
    if app.state.services:
       ^^^^^^^^^^^^^^^^^^
  File "/opt/prefect-coll/venv_prefect-coll/lib/python3.11/site-packages/starlette/datastructures.py", line 705, in __getattr__
    raise AttributeError(message.format(self.__class__.__name__, key))
AttributeError: 'State' object has no attribute 'services'
An exception occurred.
```

### Versions

```Text
Version:             2.8.7
API version:         0.8.4
Python version:      3.11.2
Git commit:          a6d6c6fc
Built:               Thu, Mar 23, 2023 3:27 PM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         <client error>
```

### Additional context

The server uses a postgresql database.

**Comment by madkinsz** [comment]
Hey @andreadistefano — glad to hear you're enjoying Prefect.

I think most of that traceback is just the server crashing. The error that looks meaningful to me is: 

> RuntimeError: Prefect requires sqlite >= 3.24.0 but we found version 3.7.17

We require a newer version of SQLite. There's an issue at #5970 that talks a bit about the best way to get a newer version. Otherwise, you could use Postgres as a backing database instead of SQLite — you'll probably want it for performance anyway.

**Comment by andreadistefano** [comment]
Hi @madkinsz, thanks for your reply! Sorry for not spotting the error, but the traceback was a bit intimidating.

I actually had already configured the Postgresql connection, and it's indeed working as expected (I see that the tables were created and there are some records inside them).

Do I still need to upgrade my SQLite version? It may not be that easy, as the server is an old CentOS 7.9, but I can try.

**Comment by madkinsz** [comment]
You shouldn't need to upgrade SQLite unless it's being used (i.e. Postgres is not being used). Where did this traceback come from?

**Comment by andreadistefano** [comment]
From `prefect work-pool ls`, but now I understood the problem: I set my Postgres connection string in the environment variable `PREFECT_API_DATABASE_CONNECTION_URL`, and the server was reading that correctly (the server is running under a systemd service).

On the other hand, when I ran the command from another terminal session, the variable was not set correctly, and that was entirely my fault. Re-setting the variable fixed it.

Sorry about this whole mess!
